photutils/psf/photometry.py

```diff
@@ -681,6 +672,21 @@ def do_photometry(self, image, init_guesses=None):
                 elif hasattr(self.psf_model, 'sigma'):
                     self.aperture_radius = (self.psf_model.sigma.value *
                                             gaussian_sigma_to_fwhm)
+                # If PSF model doesn't have FWHM or sigma value -- as it
+                # is not a Gaussian; most likely because it's an ePSF --
+                # then we fall back on fitting a circle of the average
+                # size of the fitting box. As ``fitshape`` is the width
+                # of the box, we need (width-1)/2 as the radius.
+                else:
+                    self.aperture_radius = float(np.amin((np.asanyarray(
+                                                 self.fitshape) - 1) / 2))
+                    warnings.warn('aperture_radius is None and could not '
+                                  'be determined by psf_model. Setting '
+                                  'radius to the smallest fitshape size. '
+                                  'If fitshape is significantly larger than '
+                                  'the psf_model core lengthscale, consider '
+                                  'supplying a specific aperture_radius.',
+                                  AstropyUserWarning)
```

Comment:
This will need similar logic as in `BasicPSFPhotometry'.  Currently a warning will not be raised if `init_guess` is input without a `'flux'` column.  See block starting on new line 655.  All of this logic is the same and should not be repeated code, but extracted into a helper function if possible.

Reply:
A *private* helper function.

Reply:
To add just a bit more context - @Onoddil and I discussed that probably a larger effort in this direction is called for - I made #945 to track that. But I think @larrybradley is right that we may as well start on the right foot by doing it for this specific thing, even though probably more of this is called for.

Reply:
While above in `BasicPSFPhotometry` the logic was unsound, here it is ok I believe. For iterative PSF photometry you will need `aperture_radius` set here, as in `_do_photometry` any newly detected sources will fail with the same error when trying to calculate the `aperture_flux` [at these lines](https://github.com/astropy/photutils/blob/692cdb17ed702ccd214526e0623bae7b6975fb59/photutils/psf/photometry.py#L739-L742).

These two piece of code are therefore subtly different and at this point I'm not sure you could wrap them into a helper function. Perhaps down the road once there are more iterative PSF photometry algorithms in place `BasicPSFPhotometry` can hold -- as the base class -- a helper function for _those_ cases, which would basically be this code moved out into its own function, but the iterative vs 'basic' cases here are different and should be standalone, I think.

I'm not sure how this links in with the larger issue @eteq, but I think I agree overall this function eventually should become a helper function, but `BasicPSFPhotometry` will remain an edge case which is handled specifically in its `do_photometry` call.

Reply:
Ah, I missed the subtle difference here.  In that case I think it does make sense that they should be left separate for now.  

There are some ways that one could still imagine generalizing more usefully, like having the parts *inside* the `ifs` here be wrapped up in a private function with a few parameters (and be shared in basic and iterativelysubtracted), but I think it's better to do that in a follow-on because it will probably touch a lot of places and we wouldn't want it to drag this one out further.